test/test_ion_cli.cpp

```diff
@@ -0,0 +1,567 @@
+}
+
+void test_ion_cli_add_import(const char *import, ION_CLI_IO_TYPE input_type, IonCliProcessArgs *process_args) {
+    process_args->imports.push_back(IonCliIO(import, input_type));
+}
+
+void test_ion_cli_process(const char *input, ION_CLI_IO_TYPE input_type, ION_STRING *command_output,
+                          IonEventReport *report, ION_EVENT_OUTPUT_TYPE output_format=OUTPUT_TYPE_TEXT_UGLY) {
+    IonCliCommonArgs common_args;
+    IonCliProcessArgs process_args;
+    test_ion_cli_init_common_args(&common_args, output_format);
+    test_ion_cli_add_input(input, input_type, &common_args);
+    test_ion_cli_init_process_args(&process_args);
+    ion_cli_command_process(&common_args, &process_args, command_output, report);
+}
+
+void test_ion_cli_compare(const char *input, ION_CLI_IO_TYPE input_type, ION_STRING *command_output,
+                          IonEventReport *report, ION_EVENT_COMPARISON_TYPE comparison_type,
+                          ION_EVENT_OUTPUT_TYPE output_format=OUTPUT_TYPE_TEXT_UGLY) {
+    IonCliCommonArgs common_args;
+    test_ion_cli_init_common_args(&common_args, output_format);
+    test_ion_cli_add_input(input, input_type, &common_args);
+    ion_cli_command_compare(&common_args, comparison_type, command_output, report);
+}
+
+void test_ion_cli_read_stream_successfully(BYTE *data, size_t data_len, IonEventStream *stream) {
+    hREADER reader;
+    IonEventResult result;
+    ION_ASSERT_OK(ion_test_new_reader(data, (SIZE)data_len, &reader));
+    ION_ASSERT_OK(ion_event_stream_read_all(reader, NULL, stream, &result));
+    ION_ASSERT_OK(ion_reader_close(reader));
+    ASSERT_FALSE(result.has_error_description);
+    ASSERT_FALSE(result.has_comparison_result);
+}
+
+void test_ion_cli_assert_streams_equal(const char *expected_str, ION_STRING *actual_str) {
+    IonEventStream expected("expected"), actual("actual");
+    IonEventResult result;
+
+    test_ion_cli_read_stream_successfully((BYTE *)expected_str, strlen(expected_str), &expected);
+    test_ion_cli_read_stream_successfully(actual_str->value, (size_t)actual_str->length, &actual);
+
+    ASSERT_TRUE(ion_compare_streams(&expected, &actual, &result));
+    ASSERT_FALSE(result.has_comparison_result);
+    ASSERT_FALSE(result.has_error_description);
+}
+
+TEST(IonCli, ProcessBasic) {
+    ION_STRING command_output;
+    IonEventReport report;
+    test_ion_cli_process(join_path(full_good_path, "one.ion").c_str(), IO_TYPE_FILE, &command_output, &report,
+                         OUTPUT_TYPE_EVENTS);
+    ASSERT_FALSE(report.hasComparisonFailures());
+    ASSERT_FALSE(report.hasErrors());
+    test_ion_cli_assert_streams_equal("1", &command_output);
+    test_ion_cli_assert_streams_equal("$ion_event_stream {event_type:SCALAR, ion_type:INT, value_text:\"1\", value_binary:[0xE0, 0x01, 0x00, 0xEA, 0x21, 0x01], depth: 0} {event_type: STREAM_END, depth: 0}",
+                                      &command_output);
+    free(command_output.value);
```

Comment:
The `malloc` occurs in `ion_event_writer_close`.

The tricky thing with `ION_STRING` is that it's simply a `BYTE *` and a length, and it's very widely used throughout the library. Its value is rarely actually dynamically allocated, and since it doesn't track whether it was, there isn't a good way to provide a function to clean up an `ION_STRING` every time it's about to go out of scope. And since it's so widely used, it's too late to add a flag to track how it was allocated and start requiring users to manually and unconditionally call a function to destroy it.

Turns out, this is actually the *only* place in ion-c that a dynamically-allocated `ION_STRING` is returned from a public API. Since in this case we know that the value will always be dynamically allocated, I've added an `ion_cli_free_command_output` function which takes the `ION_STRING` that was given as the output parameter to another `ion_cli_command_*` function and properly frees it.